File: insert_presentation_tags.py
#!/usr/bin/env python3

"""
========================================================
Script para insertar etiquetas de ponencias en Supabase
========================================================

Uso:
  python3 insert_presentation_tags.py tags_ponencias.json
  python3 insert_presentation_tags.py tags_ponencias.csv

Autor: Sistema de Gestión de Congreso
Fecha: 2025-11-05
"""

import sys
import json
import csv
import urllib.request
import urllib.error
from pathlib import Path

# ========================================================
# CONFIGURACIÓN DE SUPABASE
# ========================================================
SUPABASE_URL = 'https://dacpkbftkzwnpnhirgny.supabase.co'
SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImRhY3BrYmZ0a3p3bnBuaGlyZ255Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjAwMjIyMzMsImV4cCI6MjA3NTU5ODIzM30.fOIihFR2oiHSr5VWRrn5KczCZbf65Mdj7TE2kRXS6JQ'

# ========================================================
# FUNCIONES DE UTILIDAD
# ========================================================

def supabase_request(method, endpoint, data=None):
    """Hacer petición HTTP a Supabase REST API"""
    url = f'{SUPABASE_URL}/rest/v1/{endpoint}'
    headers = {
        'apikey': SUPABASE_KEY,
        'Authorization': f'Bearer {SUPABASE_KEY}',
        'Content-Type': 'application/json',
        'Prefer': 'return=representation'
    }
    
    if data:
        data = json.dumps(data).encode('utf-8')
    
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    
    try:
        with urllib.request.urlopen(req) as response:
            body = response.read().decode('utf-8')
            return json.loads(body) if body else []
    except urllib.error.HTTPError as e:
        error_body = e.read().decode('utf-8')
        raise Exception(f'HTTP {e.code}: {error_body}')

def parse_csv(file_path):
    """Leer archivo CSV y convertir a lista de diccionarios"""
    with open(file_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        return list(reader)

def insert_tags(tags):
    """Insertar etiquetas en Supabase"""
    print(f'\n📤 Insertando {len(tags)} etiquetas...')
    
    tags_to_insert = [
        {
            'name': tag['name'].lower().strip(),
            'description': tag.get('description', f"Etiqueta de ponencia: {tag.get('presentation', 'Sin especificar')}"),
            'color': tag.get('color', '#00D9C0'),
            'source': 'presentation',
            'created_by': None  # Tags de ponencias son anónimas
        }
        for tag in tags
    ]
    
    try:
        result = supabase_request('POST', 'tags', tags_to_insert)
        print(f'✅ {len(result)} etiquetas insertadas correctamente')
        return result
    except Exception as error:
        # Si hay duplicados, intentar uno por uno
        if '23505' in str(error) or 'duplicate' in str(error).lower():
            print('⚠️  Detectados duplicados. Insertando uno por uno...')
            inserted = 0
            skipped = 0
            errors = 0
            
            for tag in tags_to_insert:
                try:
                    supabase_request('POST', 'tags', [tag])
                    inserted += 1
                    print(f"  ✓ {tag['name']}")
                except Exception as err:
                    if '23505' in str(err) or 'duplicate' in str(err).lower():
                        skipped += 1
                        print(f"  ⊘ {tag['name']} (ya existe)")
                    else:
                        errors += 1
                        print(f"  ✗ {tag['name']}: {err}")
            
            print(f'\n📊 Resultado:')
            print(f'   ✅ Insertadas: {inserted}')
            print(f'   ⊘ Omitidas (duplicadas): {skipped}')
            print(f'   ✗ Errores: {errors}')
            
            return {'inserted': inserted, 'skipped': skipped, 'errors': errors}
        else:
            raise

def show_tags_summary():
    """Mostrar resumen de etiquetas existentes"""
    try:
        tags = supabase_request('GET', 'tags?select=source&order=created_at.desc')
        
        by_source = {}
        for tag in tags:
            source = tag.get('source', 'user')
            by_source[source] = by_source.get(source, 0) + 1
        
        print('\n📊 Resumen de etiquetas en BD:')
        print(f"   Total: {len(tags)}")
        print(f"   - Usuario: {by_source.get('user', 0)}")
        print(f"   - Ponencias: {by_source.get('presentation', 0)}")
    except Exception as error:
        print(f'Error al obtener resumen: {error}')

# ========================================================
# MAIN
# ========================================================

def main():
    print('=' * 56)
    print('📝 Script de Inserción de Etiquetas de Ponencias')
    print('=' * 56)
    print()
    
    # Validar argumentos
    if len(sys.argv) < 2:
        print('❌ Error: Debes proporcionar un archivo de entrada\n')
        print('Uso:')
        print('  python3 insert_presentation_tags.py tags_ponencias.json')
        print('  python3 insert_presentation_tags.py tags_ponencias.csv\n')
        sys.exit(1)
    
    file_path = sys.argv[1]
    
    # Verificar que el archivo existe
    if not Path(file_path).exists():
        print(f'❌ Error: El archivo "{file_path}" no existe\n')
        sys.exit(1)
    
    try:
        # Leer archivo según extensión
        ext = Path(file_path).suffix.lower()
        
        if ext == '.json':
            print(f'📂 Leyendo archivo JSON: {file_path}')
            with open(file_path, 'r', encoding='utf-8') as f:
                tags = json.load(f)
        elif ext == '.csv':
            print(f'📂 Leyendo archivo CSV: {file_path}')
            tags = parse_csv(file_path)
        else:
            print(f'❌ Error: Formato de archivo no soportado ({ext})')
            print('Formatos válidos: .json, .csv\n')
            sys.exit(1)
        
        # Validar estructura
        if not isinstance(tags, list) or len(tags) == 0:
            print('❌ Error: El archivo debe contener un array con al menos una etiqueta\n')
            sys.exit(1)
        
        print(f'✅ {len(tags)} etiquetas leídas correctamente\n')
        
        # Mostrar preview
        print('👁️  Preview de etiquetas a insertar:')
        for i, tag in enumerate(tags[:5], 1):
            presentation = tag.get('presentation', 'sin ponencia')
            print(f"   {i}. {tag['name']} ({presentation})")
        if len(tags) > 5:
            print(f'   ... y {len(tags) - 5} más\n')
        
        # Insertar en Supabase
        insert_tags(tags)
        
        # Mostrar resumen final
        show_tags_summary()
        
        print('\n✅ Proceso completado exitosamente\n')
        print('=' * 56)
        print()
        
    except Exception as error:
        print(f'\n❌ Error fatal: {error}')
        print(f'\nDetalles: {error.__class__.__name__}')
        sys.exit(1)

if __name__ == '__main__':
    main()

File: insert_presentation_tags.js
#!/usr/bin/env node

/**
 * ========================================================
 * Script para insertar etiquetas de ponencias en Supabase
 * ========================================================
 * 
 * Uso:
 *   node insert_presentation_tags.js tags_ponencias.json
 *   node insert_presentation_tags.js tags_ponencias.csv
 * 
 * Autor: Sistema de Gestión de Congreso
 * Fecha: 2025-11-05
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

// ========================================================
// CONFIGURACIÓN DE SUPABASE
// ========================================================
const SUPABASE_URL = 'https://dacpkbftkzwnpnhirgny.supabase.co';
const SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImRhY3BrYmZ0a3p3bnBuaGlyZ255Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjAwMjIyMzMsImV4cCI6MjA3NTU5ODIzM30.fOIihFR2oiHSr5VWRrn5KczCZbf65Mdj7TE2kRXS6JQ';

// ========================================================
// FUNCIONES DE UTILIDAD
// ========================================================

/**
 * Hacer petición HTTP a Supabase REST API
 */
function supabaseRequest(method, endpoint, data = null) {
    return new Promise((resolve, reject) => {
        const options = {
            hostname: 'dacpkbftkzwnpnhirgny.supabase.co',
            path: `/rest/v1/${endpoint}`,
            method: method,
            headers: {
                'apikey': SUPABASE_KEY,
                'Authorization': `Bearer ${SUPABASE_KEY}`,
                'Content-Type': 'application/json',
                'Prefer': 'return=representation'
            }
        };

        const req = https.request(options, (res) => {
            let body = '';
            res.on('data', (chunk) => body += chunk);
            res.on('end', () => {
                if (res.statusCode >= 200 && res.statusCode < 300) {
                    resolve(JSON.parse(body || '[]'));
                } else {
                    reject(new Error(`HTTP ${res.statusCode}: ${body}`));
                }
            });
        });

        req.on('error', reject);
        if (data) req.write(JSON.stringify(data));
        req.end();
    });
}

/**
 * Leer archivo CSV y convertir a JSON
 */
function parseCSV(filePath) {
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.trim().split('\n');
    const headers = lines[0].split(',').map(h => h.trim());
    
    return lines.slice(1).map(line => {
        const values = line.split(',').map(v => v.trim());
        const obj = {};
        headers.forEach((header, i) => {
            obj[header] = values[i] || '';
        });
        return obj;
    });
}

/**
 * Insertar etiquetas en Supabase
 */
async function insertTags(tags) {
    console.log(`\n📤 Insertando ${tags.length} etiquetas...`);
    
    const tagsToInsert = tags.map(tag => ({
        name: tag.name.toLowerCase().trim(),
        description: tag.description || `Etiqueta de ponencia: ${tag.presentation || 'Sin especificar'}`,
        color: tag.color || '#00D9C0',
        source: 'presentation',
        created_by: null // Tags de ponencias son anónimas
    }));

    try {
        const result = await supabaseRequest('POST', 'tags', tagsToInsert);
        console.log(`✅ ${result.length} etiquetas insertadas correctamente`);
        return result;
    } catch (error) {
        // Si hay duplicados (error 23505), intentar uno por uno
        if (error.message.includes('23505') || error.message.includes('duplicate')) {
            console.log('⚠️  Detectados duplicados. Insertando uno por uno...');
            let inserted = 0;
            let skipped = 0;
            
            for (const tag of tagsToInsert) {
                try {
                    await supabaseRequest('POST', 'tags', [tag]);
                    inserted++;
                    console.log(`  ✓ ${tag.name}`);
                } catch (err) {
                    if (err.message.includes('23505') || err.message.includes('duplicate')) {
                        skipped++;
                        console.log(`  ⊘ ${tag.name} (ya existe)`);
                    } else {
                        console.error(`  ✗ ${tag.name}: ${err.message}`);
                    }
                }
            }
            
            console.log(`\n📊 Resultado:`);
            console.log(`   ✅ Insertadas: ${inserted}`);
            console.log(`   ⊘ Omitidas (duplicadas): ${skipped}`);
            console.log(`   ✗ Errores: ${tagsToInsert.length - inserted - skipped}`);
            
            return { inserted, skipped };
        } else {
            throw error;
        }
    }
}

/**
 * Mostrar resumen de etiquetas existentes
 */
async function showTagsSummary() {
    try {
        const tags = await supabaseRequest('GET', 'tags?select=source&order=created_at.desc');
        
        const bySource = tags.reduce((acc, tag) => {
            acc[tag.source || 'user'] = (acc[tag.source || 'user'] || 0) + 1;
            return acc;
        }, {});
        
        console.log('\n📊 Resumen de etiquetas en BD:');
        console.log(`   Total: ${tags.length}`);
        console.log(`   - Usuario: ${bySource.user || 0}`);
        console.log(`   - Ponencias: ${bySource.presentation || 0}`);
    } catch (error) {
        console.error('Error al obtener resumen:', error.message);
    }
}

// ========================================================
// MAIN
// ========================================================

async function main() {
    console.log('========================================================');
    console.log('📝 Script de Inserción de Etiquetas de Ponencias');
    console.log('========================================================\n');

    // Validar argumentos
    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.error('❌ Error: Debes proporcionar un archivo de entrada\n');
        console.log('Uso:');
        console.log('  node insert_presentation_tags.js tags_ponencias.json');
        console.log('  node insert_presentation_tags.js tags_ponencias.csv\n');
        process.exit(1);
    }

    const filePath = args[0];
    
    // Verificar que el archivo existe
    if (!fs.existsSync(filePath)) {
        console.error(`❌ Error: El archivo "${filePath}" no existe\n`);
        process.exit(1);
    }

    try {
        // Leer archivo según extensión
        let tags;
        const ext = path.extname(filePath).toLowerCase();
        
        if (ext === '.json') {
            console.log(`📂 Leyendo archivo JSON: ${filePath}`);
            const content = fs.readFileSync(filePath, 'utf-8');
            tags = JSON.parse(content);
        } else if (ext === '.csv') {
            console.log(`📂 Leyendo archivo CSV: ${filePath}`);
            tags = parseCSV(filePath);
        } else {
            console.error(`❌ Error: Formato de archivo no soportado (${ext})`);
            console.log('Formatos válidos: .json, .csv\n');
            process.exit(1);
        }

        // Validar estructura
        if (!Array.isArray(tags) || tags.length === 0) {
            console.error('❌ Error: El archivo debe contener un array con al menos una etiqueta\n');
            process.exit(1);
        }

        console.log(`✅ ${tags.length} etiquetas leídas correctamente\n`);

        // Mostrar preview
        console.log('👁️  Preview de etiquetas a insertar:');
        tags.slice(0, 5).forEach((tag, i) => {
            console.log(`   ${i + 1}. ${tag.name} (${tag.presentation || 'sin ponencia'})`);
        });
        if (tags.length > 5) {
            console.log(`   ... y ${tags.length - 5} más\n`);
        }

        // Insertar en Supabase
        await insertTags(tags);

        // Mostrar resumen final
        await showTagsSummary();

        console.log('\n✅ Proceso completado exitosamente\n');
        console.log('========================================================\n');

    } catch (error) {
        console.error('\n❌ Error fatal:', error.message);
        console.error('\nDetalles:', error);
        process.exit(1);
    }
}

// Ejecutar
main();
